# **Tratamento dos dados de Projetos de Monitoramento de Praia (PMPs) entre os anos de 2020-2024**

## **Descrição**
### **Objetivo**:

Este notebook realiza a limpeza e o tratamento dos dados para análises futuras no Power BI, garantindo melhor desempenho das consultas e visualizações no dashboard.

### **Dados**:

Os arquivos em CSV foram obtidos do Sistema de Informação de Monitoramento da Biota Aquática ([SIMBA](https://simba.petrobras.com.br/simba/web/)), utilizando filtro de datas de 01/01/2020 a 31/12/2024. Após o download, os arquivos foram hospedados em Google Drive com compartilhamento público (links disponíveis em [utils.py](https://github.com/tainahguerras/PMP-Petrobras-Analise/blob/main/utils.py)).

### **Etapas**:
1. Importação das Bibliotecas

2. Importação dos Dados

3. Inspeção dos Dados

4. Transformação dos Dados

5. Exportação do Dataset Tratado


## **Importação de Bibliotecas**

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)

import utils as ut # importa listas e funções definidas em utils.py

## **Importação dos Dados**

Durante a importação dos dados, definimos corretamente o tipo da coluna "Data/Hora", garantindo que o formato DD/MM/AAAA fosse reconhecido. Além disso, selecionamos previamente um conjunto de colunas relevantes para inspeção neste notebook.

In [2]:
datasets = {}
nomes = ['BC_ES', 'BS_RJ', 'BS_SC_PR', 'BS_SP', 'RN_CE', 'SE_AL']

urls_convertidas = [ut.conversor_url(url) for url in ut.urls] # convertendo urls de compartilhamento do sheets para formato aceito por .read_csv()

# Criando datasets e armazenando em dicionário
for i in range(len(nomes)): 
    datasets[nomes[i]] = pd.read_csv(urls_convertidas[i],parse_dates=['Data/Hora'],dayfirst=True,low_memory=False)[ut.colunas_selecionadas]

## **Inspeção dos Dados**

In [3]:
datasets['SE_AL'].head() # Exemplo de dataset criado

Código  Identificador do indivíduo Instituição executora   Estado  \
0  325581                      136459                 Biota  Alagoas   
1  325580                      136240                 Biota  Alagoas   
2  325489                      131188                 TAMAR  Sergipe   
3  325456                      136554                 TAMAR    Bahia   
4  325410                      137100                 TAMAR  Sergipe   

                          Cidade                                    Praia  \
0                      Piaçabuçu     zz desativado 1 Praia Pontal do Peba   
1                      Piaçabuçu     zz desativado 1 Praia Pontal do Peba   
2                        Aracaju            zz desativado 1 Praia Aracaju   
3                       Jandaíra        zz desativado 1 Praia Mangue Seco   
4  Brejo Grande,Pacatuba,Pirambu  zz desativado 1 Praia Ponta dos Mangues   

                           Trecho Estratégia do trecho Maré inicial  \
0      zz desativado 1 APA (Peba)               Diário      Vazante   
1      zz desativado 1 APA (Peba)               Diário      Vazante   
2         zz desativado 1 Aracaju               Diário     Enchente   
3     zz desativado 1 Mangue Seco               Diário     Enchente   
4  zz desativado 1 Anuência_Norte               Diário     Enchente   

                      Vento inicial Tipo do monitoramento           Data/Hora  \
0    2 - Brisa leve – 6 até 11 km/h               Regular 2024-10-22 11:44:00   
1    2 - Brisa leve – 6 até 11 km/h               Regular 2024-10-22 11:24:00   
2  4 - Brisa fraca – 20 até 29 km/h               Regular 2024-10-22 09:39:00   
3    2 - Brisa leve – 6 até 11 km/h               Regular 2024-10-22 07:15:00   
4    2 - Brisa leve – 6 até 11 km/h               Regular 2024-10-22 06:25:00   

    Ponto - Lat  Ponto - Long Espécies - Classe Espécies - Ordem  \
0    -1.038.249  -363.215.983          Reptilia       Testudines   
1  -103.717.217  -363.134.683          Reptilia       Testudines   
2  -110.581.917    -3.709.627          Reptilia       Testudines   
3  -114.830.733  -373.694.517          Reptilia       Testudines   
4      -105.369       -36.494          Reptilia       Testudines   

  Espécies - Subordem Espécies - Família Espécies - Gênero  \
0          Cryptodira        Cheloniidae      Lepidochelys   
1          Cryptodira        Cheloniidae      Lepidochelys   
2          Cryptodira        Cheloniidae          Chelonia   
3          Cryptodira        Cheloniidae      Lepidochelys   
4          Cryptodira        Cheloniidae          Chelonia   

      Espécies - Espécie Espécie ameaçada Caracterização do ambiente Condição  
0  Lepidochelys olivacea              Sim                    Arenoso    Morto  
1  Lepidochelys olivacea              Sim                    Arenoso    Morto  
2         Chelonia mydas              Não                    Arenoso    Morto  
3  Lepidochelys olivacea              Sim                    Arenoso    Morto  
4         Chelonia mydas              Não                    Arenoso    Morto

**Checando tipagem dos dados**

In [4]:
datasets['SE_AL'].dtypes

Código                                 int64
Identificador do indivíduo             int64
Instituição executora                 object
Estado                                object
Cidade                                object
Praia                                 object
Trecho                                object
Estratégia do trecho                  object
Maré inicial                          object
Vento inicial                         object
Tipo do monitoramento                 object
Data/Hora                     datetime64[ns]
Ponto - Lat                           object
Ponto - Long                          object
Espécies - Classe                     object
Espécies - Ordem                      object
Espécies - Subordem                   object
Espécies - Família                    object
Espécies - Gênero                     object
Espécies - Espécie                    object
Espécie ameaçada                      object
Caracterização do ambiente            object
Condição  

**Checando filtragem das datas**

Vamos verificar se a filtragem aplicada no SIMBA para o download dos datasets foi correta, considerando o intervalo definido: 01/01/2020 a 31/12/2024.

In [5]:
for nome, df in datasets.items():
    print(f"Dataset: {nome}")
    print(f"  Data mínima: {df['Data/Hora'].min()}")
    print(f'  Data máxima: {df['Data/Hora'].max()}')
    print('-' * 27)

Dataset: BC_ES
  Data mínima: 2020-01-01 07:16:00
  Data máxima: 2024-11-03 11:36:00
---------------------------
Dataset: BS_RJ
  Data mínima: 2020-01-01 06:49:00
  Data máxima: 2024-11-01 08:09:00
---------------------------
Dataset: BS_SC_PR
  Data mínima: 2020-01-01 08:15:00
  Data máxima: 2024-11-02 08:28:00
---------------------------
Dataset: BS_SP
  Data mínima: 2020-01-01 05:27:00
  Data máxima: 2024-11-03 08:14:00
---------------------------
Dataset: RN_CE
  Data mínima: 2020-01-02 07:21:00
  Data máxima: 2024-07-31 09:06:00
---------------------------
Dataset: SE_AL
  Data mínima: 2020-01-01 12:46:00
  Data máxima: 2024-10-22 11:44:00
---------------------------


**Checando tamanhos dos datasets**

In [6]:
for nome, df in datasets.items():
    print(f'{nome}: {df.shape}')

BC_ES: (24246, 23)
BS_RJ: (10565, 23)
BS_SC_PR: (51776, 23)
BS_SP: (23576, 23)
RN_CE: (3488, 23)
SE_AL: (9246, 23)


**Checando registros duplicados**

In [7]:
for nome, df in datasets.items():
    print(f'{nome}: {df.duplicated().sum()} registros duplicados')

BC_ES: 0 registros duplicados
BS_RJ: 0 registros duplicados
BS_SC_PR: 0 registros duplicados
BS_SP: 0 registros duplicados
RN_CE: 0 registros duplicados
SE_AL: 0 registros duplicados


**Checando registros ausentes**

In [8]:
for nome, df in datasets.items():
    print(f'Dataset: {nome}\n')
    print(df.isna().sum()[df.isna().sum() > 0])
    print('-' * 33)

Dataset: BC_ES

Maré inicial                1
Vento inicial               1
Tipo do monitoramento      99
Espécies - Subordem      4410
Espécies - Família         42
Espécies - Gênero          78
Espécies - Espécie        140
Espécie ameaçada          140
dtype: int64
---------------------------------
Dataset: BS_RJ

Maré inicial                4
Vento inicial               4
Tipo do monitoramento      13
Espécies - Subordem      5931
Espécies - Família         47
Espécies - Gênero          98
Espécies - Espécie        363
Espécie ameaçada          363
dtype: int64
---------------------------------
Dataset: BS_SC_PR

Espécies - Ordem           2
Espécies - Subordem    37782
Espécies - Família       104
Espécies - Gênero        230
Espécies - Espécie       668
Espécie ameaçada         668
dtype: int64
---------------------------------
Dataset: BS_SP

Maré inicial                13
Vento inicial               13
Tipo do monitoramento      148
Espécies - Subordem      12251
Espécies - Fam

Dentre as colunas que apresentaram alta quantidade de registros nulos, a de "Espécie ameaçada" é a que trás mais relevância para as análises, pois possibilita discussão acerca de impactos em espécies com maior vulnerabilidade. 

Para decidir se vale a pena mantê-la nos dados, vamos analisar mais proximamente a proporção de registros ausentes em relação ao total.

In [9]:
for nome, df in datasets.items():
    ut.calcular_porcentagem_nulos(df, nome, 'Espécie ameaçada')

BC_ES: 0.58% de valores ausentes
BS_RJ: 3.44% de valores ausentes
BS_SC_PR: 1.29% de valores ausentes
BS_SP: 2.80% de valores ausentes
RN_CE: 3.01% de valores ausentes
SE_AL: 8.17% de valores ausentes


A maioria dos datasets apresentou baixa taxa de valores ausentes na coluna "Espécie Ameaçada", com SE_AL tendo o maior percentual (8.17%). Esse valor ainda é aceitável e não compromete as análises, justificando a permanência da coluna.

## **Tranformação dos Dados**

Nesta etapa, foram realizadas as seguintes operações:

* Concatenação dos datasets em um único arquivo consolidado
* Remoção de colunas irrelevantes para os objetivos da análise
* Ajuste na coluna "Data/Hora", mantendo apenas a data (sem a informação de horário)
* Padronização da coluna "Cidade", registrando apenas a primeira cidade mencionada quando múltiplas foram inseridas

**Concatenando datasets**

In [10]:
mapeamento_regioes = {
    'BC_ES': 'Bacias de Campos e Espírito Santo',
    'BS_RJ': 'Bacia de Santos',
    'BS_SC_PR': 'Bacia de Santos',
    'BS_SP': 'Bacia de Santos',
    'RN_CE': 'Bacia Potiguar',
    'SE_AL': 'Bacia de Sergipe-Alagoas'
}

# Adicionando a nova coluna "Região" a cada dataset
for nome, df in datasets.items():
    df['Região'] = mapeamento_regioes[nome]

# Concatenando todos os datasets em um só
df_concat = pd.concat(datasets.values(), ignore_index=True)
    

**Tratando colunas**

In [11]:
colunas_dispensaveis = ['Identificador do indivíduo', 'Trecho', 'Maré inicial', 'Praia', 'Vento inicial', 'Tipo do monitoramento', 'Espécies - Subordem', 'Espécies - Família', 'Espécies - Gênero']

# Remove colunas dispensáveis
df_tratado = df_concat.drop(columns=colunas_dispensaveis).dropna().copy() 

# Tratando coluna "Data/Hora"
df_tratado['Data/Hora'] = df_tratado['Data/Hora'].dt.date    # remove horários da coluna "Data/Hora"
df_tratado.rename(columns={'Data/Hora':'Data'},inplace=True) # muda nome da coluna "Data/Hora" para "Data"

# Tratando coluna "Cidade"
df_tratado = df_tratado.join(df_tratado['Cidade'].str.split(',', expand=True).add_prefix('Cidade ')) # divide coluna 'Cidade", com múltiplos registros separados por vírgula, em múltiplas colunas
colunas_drop = [col for col in df_tratado.columns if col.startswith('Cidade') and col != 'Cidade 0'] # define dinamicamente colunas a serem deletadas, exceto a contendo somente a primeira cidade
df_tratado = df_tratado.drop(columns=colunas_drop)                                                   # remove colunas extras de Cidade
df_tratado.rename(columns={'Cidade 0':'Cidade'},inplace=True)                                        # renomeia coluna de cidade

**Checando modificações**

In [12]:
df_tratado.head()

Código                              Instituição executora          Estado  \
0  327574  Instituto de Pesquisa e Conservação Marinha - ...  Espírito Santo   
1  327572  Instituto de Pesquisa e Conservação Marinha - ...  Espírito Santo   
2  327564  Instituto de Pesquisa e Conservação Marinha - ...  Espírito Santo   
3  327563  Instituto de Pesquisa e Conservação Marinha - ...  Espírito Santo   
4  327562  Instituto de Pesquisa e Conservação Marinha - ...  Espírito Santo   

  Estratégia do trecho        Data Ponto - Lat Ponto - Long Espécies - Classe  \
0               Diário  2024-11-03    -210.369     -408.178          Reptilia   
1               Diário  2024-11-03    -210.963     -408.462          Reptilia   
2               Diário  2024-11-03    -210.583     -408.333          Reptilia   
3               Diário  2024-11-03    -209.304     -407.981          Reptilia   
4               Diário  2024-11-03    -211.387     -408.763          Reptilia   

  Espécies - Ordem Espécies - Espécie Espécie ameaçada  \
0       Testudines    Caretta caretta              Não   
1       Testudines     Chelonia mydas              Não   
2       Testudines     Chelonia mydas              Não   
3       Testudines     Chelonia mydas              Não   
4       Testudines    Caretta caretta              Não   

  Caracterização do ambiente Condição                             Região  \
0                    Arenoso    Morto  Bacias de Campos e Espírito Santo   
1                    Arenoso    Morto  Bacias de Campos e Espírito Santo   
2                    Arenoso    Morto  Bacias de Campos e Espírito Santo   
3                    Arenoso    Morto  Bacias de Campos e Espírito Santo   
4                    Arenoso    Morto  Bacias de Campos e Espírito Santo   

       Cidade  
0  Marataízes  
1  Marataízes  
2  Marataízes  
3  Itapemirim  
4  Marataízes

In [13]:
# Checando deleção de registros com valores ausentes
if df_tratado.isna().sum().sum() == 0: 
    print("Não há valores ausentes.")
else:
    print("Valores ausentes encontrados.")

# Checando tamanho final do dataset
print(f'Tamanho final do dataset: {df_tratado.shape}')

Não há valores ausentes.
Tamanho final do dataset: (120068, 15)


In [14]:
# Checando nova coluna "Região"

print('Quantidade de registros por região:')
df_tratado['Região'].value_counts()

Quantidade de registros por região:


Região
Bacia de Santos                      84225
Bacias de Campos e Espírito Santo    24106
Bacia de Sergipe-Alagoas              8491
Bacia Potiguar                        3246
Name: count, dtype: int64

## **Exportação do Dataset Tratado**

O dataset tratado foi convertido para o formato CSV e salvo em [dataset_tratado](https://github.com/tainahguerras/PMP-Petrobras-Analise/tree/main/dataset_tratado). Ele será utilizado posteriormente na criação do dashboard no Power BI.

In [15]:
import os

# Criando pasta e fazendo download de dataset tratado em .csv 

pasta_saida = "dataset_tratado"

os.makedirs("dataset_tratado", exist_ok=True) # cria pasta de saída no repositório

caminho_arquivo = os.path.join(pasta_saida, "PMP_dados.csv")
df_tratado.to_csv(caminho_arquivo, index=False, encoding="utf-8")